variable/include/scipp/variable/shape.h

```diff
@@ -41,4 +46,18 @@ SCIPP_VARIABLE_EXPORT void squeeze(Variable &var, const std::vector<Dim> &dims);
 
 SCIPP_VARIABLE_EXPORT void expect_same_volume(const Dimensions &old_dims,
                                               const Dimensions &new_dims);
+
+SCIPP_VARIABLE_EXPORT void validate_fold_dims(const Dimensions &old_dims,
+                                              const Dim from_dim,
+                                              const Dimensions &to_dims);
+SCIPP_VARIABLE_EXPORT
+void validate_flatten_dims(const Dimensions &old_dims,
+                           const Dimensions &from_dims, const Dim to_dim);
+SCIPP_VARIABLE_EXPORT Dimensions fold_dims(const Dimensions &old_dims,
+                                           const Dim from_dim,
+                                           const Dimensions &to_dims);
+SCIPP_VARIABLE_EXPORT Dimensions flatten_dims(const Dimensions &old_dims,
+                                              const Dimensions &from_dims,
+                                              const Dim to_dim);
```

Comment:
- `fold_dims` and `flatten_dims` probably better belong into `core/`, maybe `dimensions.h`.
- It is odd that there are separate validation functions, shouldn't this be part of the actual functions?
- I would name these `fold` and `flatten`, dropping the `_dims`, just like we have other functions for `Dimensions`.

Reply:
The separate validation functions were so that the validation could be performed only once, instead of inside every coord and mask in addition to the data.
Also, the check on, for example, whether the variable dims contain the dim to be folded, are a bit different when you are dealing with data array coords.
```C++
if (!old_dims.contains(from_dim))
    throw except::DimensionError("Dimension to fold not found.");
```
is valid for the `data` of a data array, because that data variable is supposed to contain the dim to be folded, but if you have a 1d `x` coordinate on a 2d data array, and you are folding `y`, then this check throws, when it shouldn't.

I guess this is probably an indication that I should write different checks for folding a Variable and a DataArray

Reply:
> The separate validation functions were so that the validation could be performed only once, instead of inside every coord and mask in addition to the data.

I don't think this would be a performance issue. More importantly:

You are declaring functions as part of our public API. They should therefore do the appropriate validation of inputs. If it is an implementation detail, keep it in the `.cpp`.

Reply:
yes i wanted to keep it in the cpp, but i couldn't find a good way to share it between Variable and Dataset. Maybe you have an example of how you do this?

Reply:
Maybe we can first figure out why we can't turn this into proper `fold` and `flatten` for `class Dimensions`?

Reply:
But to answer your question: You'd put it into a header that is not in `include/scipp/module`, but a header alongside the `cpp` files.

Reply:
Yes indeed this would be wrong. Not necessarily strange, depending on how you create a masks it can easily end up transposed, compare:
```python
data.masks['mask1'] = x < y
data.masks['mask2'] = y > x
```